docs/source/extend/function-groups.md

```diff
@@ -55,181 +141,843 @@ from nat.builder.function import FunctionGroup
-    include: list[str] = ["func1", "func2", "func3"]
-    environment: str = "development"
+from collections.abc import Sequence
+
+class ComplexFilterConfig(FunctionGroupBaseConfig, name="complex_filter_group"):
+    include: list[str] = Field(default_factory=lambda: ["func1", "func2", "func3", "test_func4"])
+    environment: str = Field(default="development")
+    enable_experimental: bool = Field(default=False)
 
-@register_function_group(config_type=FilteredGroupConfig)
-async def build_filtered_group(config: FilteredGroupConfig, _builder: Builder):
-    # Group filter: only production-ready functions in production
+@register_function_group(config_type=ComplexFilterConfig)
+async def build_complex_group(config: ComplexFilterConfig, _builder: Builder):
+    # Group-level filter: Remove test functions in production
     async def env_filter(names: Sequence[str]) -> Sequence[str]:
         if config.environment == "production":
             return [name for name in names if not name.startswith("test_")]
         return names
     
-    # Per-function filter: exclude experimental features
-    async def stable_only(name: str) -> bool:
-        return not name.endswith("_experimental")
+    # Per-function filter: Only include experimental if flag is set
+    async def experimental_gate(name: str) -> bool:
+        return config.enable_experimental
     
     group = FunctionGroup(config=config, filter_fn=env_filter)
     
-    group.add_function("func1", fn1)  # Included by config
-    group.add_function("test_func2", fn2)  # Included by config, but filtered by env_filter in production
-    group.add_function("func3_experimental", fn3, filter_fn=stable_only)  # Excluded by per-function filter
+    # Add functions
+    group.add_function("func1", fn1)           # Always included
+    group.add_function("func2", fn2)           # Always included
+    group.add_function("func3_experimental",   # Conditionally included
+                      fn3, 
+                      filter_fn=experimental_gate)
+    group.add_function("test_func4", fn4)     # Removed in production by group filter
+    group.add_function("func5", fn5)          # NOT in include list, so never accessible
     
     yield group
 ```
 
+**Result in production** (`environment="production"`, `enable_experimental=False`):
+1. Start with: `["func1", "func2", "func3_experimental", "test_func4"]` (include list)
+2. After group filter: `["func1", "func2", "func3_experimental"]` (test_func4 removed)
+3. After per-function filter: `["func1", "func2"]` (func3_experimental removed)
+
+**Result in development** (`environment="development"`, `enable_experimental=True`):
+1. Start with: `["func1", "func2", "func3_experimental", "test_func4"]` (include list)
+2. After group filter: `["func1", "func2", "func3_experimental", "test_func4"]` (all pass)
+3. After per-function filter: `["func1", "func2", "func3_experimental", "test_func4"]` (all pass)
+
 ## Best Practices
```

Comment:
A lot of this seems duplicated with the workflows section. Can we deduplicate and have one best practices section and link?

Reply:
Removing from here